## Data preparation HW
### Group: 22_LR_JA
### Author: Yauheni Chekan

In [98]:
import warnings

import pandas as pd
from sklearn.preprocessing import MinMaxScaler, RobustScaler

pd.set_option("display.float_format", "{:.4f}".format)

warnings.filterwarnings("ignore")

In [99]:
df = pd.read_csv("datasets/Building_Permits.csv")

In [100]:
df.head(5)

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           05/06/2015  0326  023            140                  NaN   
1           04/19/2016  0306  007            440                  NaN   
2           05/27/2016  0595  203           1647                  NaN   
3           11/07/2016  0156  011           1230                  NaN   
4           11/28/2016  0342  001            950                  NaN   

  Street Name Street Suffix  ...  Existing Construction Type  \
0       Ellis            St  ...                      3.0000   
1       Geary            St  ...                      3.0000   
2     Pacific            Av  ...                      1.0000   
3     Pacific            Av  ...                      5.0000   
4      Market            St  ...                      3.0000   

  Existing Construction Type Description Proposed Construction Type  \
0                          constr type 3                        NaN   
1                          constr type 3                        NaN   
2                          constr type 1                     1.0000   
3                         wood frame (5)                     5.0000   
4                          constr type 3                        NaN   

  Proposed Construction Type Description Site Permit Supervisor District  \
0                                    NaN         NaN              3.0000   
1                                    NaN         NaN              3.0000   
2                          constr type 1         NaN              3.0000   
3                         wood frame (5)         NaN              3.0000   
4                                    NaN         NaN              6.0000   

  Neighborhoods - Analysis Boundaries    Zipcode  \
0                          Tenderloin 94102.0000   
1                          Tenderloin 94102.0000   
2                        Russian Hill 94109.0000   
3                            Nob Hill 94109.0000   
4                          Tenderloin 94102.0000   

                                    Location      Record ID  
0  (37.785719256680785, -122.40852313194863)  1380611233945  
1   (37.78733980600732, -122.41063199757738)  1420164406718  
2    (37.7946573324287, -122.42232562979227)  1424856504716  
3   (37.79595867909168, -122.41557405519474)  1443574295566  
4   (37.78315261897309, -122.40950883997789)   144548169992  

[5 rows x 43 columns]

### Describe the dataframe:

In [101]:
df.describe().T

count               mean               std  \
Permit Type                198900.0000             7.5223            1.4575   
Street Number              198900.0000          1121.7289         1135.7689   
Unit                        29479.0000            78.5172          326.9813   
Number of Existing Stories 156116.0000             5.7058            8.6135   
Number of Proposed Stories 156032.0000             5.7450            8.6133   
Estimated Cost             160834.0000        168955.4433      3630385.9537   
Revised Cost               192834.0000        132856.1865      3584902.5920   
Existing Units             147362.0000            15.6662           74.4763   
Proposed Units             147989.0000            16.5110           75.2204   
Plansets                   161591.0000             1.2747           22.4073   
Existing Construction Type 155534.0000             4.0729            1.5858   
Proposed Construction Type 155738.0000             4.0895            1.5788   
Supervisor District        197183.0000             5.5384            2.8870   
Zipcode                    197184.0000         94115.5006            9.2701   
Record ID                  198900.0000 1162047577771.6997 491821537303.4352   

                                        min                25%  \
Permit Type                          1.0000             8.0000   
Street Number                        0.0000           235.0000   
Unit                                 0.0000             0.0000   
Number of Existing Stories           0.0000             2.0000   
Number of Proposed Stories           0.0000             2.0000   
Estimated Cost                       1.0000          3300.0000   
Revised Cost                         0.0000             1.0000   
Existing Units                       0.0000             1.0000   
Proposed Units                       0.0000             1.0000   
Plansets                             0.0000             0.0000   
Existing Construction Type           1.0000             3.0000   
Proposed Construction Type           1.0000             3.0000   
Supervisor District                  1.0000             3.0000   
Zipcode                          94102.0000         94109.0000   
Record ID                  12935321500.0000 1308566734866.2500   

                                          50%                75%  \
Permit Type                            8.0000             8.0000   
Street Number                        710.0000          1700.0000   
Unit                                   0.0000             1.0000   
Number of Existing Stories             3.0000             4.0000   
Number of Proposed Stories             3.0000             4.0000   
Estimated Cost                     11000.0000         35000.0000   
Revised Cost                        7000.0000         28707.5000   
Existing Units                         1.0000             4.0000   
Proposed Units                         2.0000             4.0000   
Plansets                               2.0000             2.0000   
Existing Construction Type             5.0000             5.0000   
Proposed Construction Type             5.0000             5.0000   
Supervisor District                    6.0000             8.0000   
Zipcode                            94114.0000         94122.0000   
Record ID                  1371839772769.5000 1434999616362.5000   

                                          max  
Permit Type                            8.0000  
Street Number                       8400.0000  
Unit                                6004.0000  
Number of Existing Stories            78.0000  
Number of Proposed Stories            78.0000  
Estimated Cost                 537958646.0000  
Revised Cost                   780500000.0000  
Existing Units                      1907.0000  
Proposed Units                      1911.0000  
Plansets                            9000.0000  
Existing Construction Type             5.0000  
Proposed Construction Type             5.0000 

### Missing values for "Existing Use" column can be populated with the mode.

Mode for 'Existing Use' column is `1 family dwelling`

In [102]:
print(f"Mode for 'Existing Use' column: '{df['Existing Use'].mode().to_numpy()[0]}'")


Mode for 'Existing Use' column: '1 family dwelling'


### Define the amount of missing values in "Current Status" field:

Amount of missing values in 'Current Status' field: 0 (0.00%)

In [103]:
missing_values = df["Current Status"].isna().sum()
missing_percentage = (missing_values / len(df)) * 100
print(f"Amount of missing values in 'Current Status' field: {missing_values} ({missing_percentage:.2f}%)")


Amount of missing values in 'Current Status' field: 0 (0.00%)


### Boxplot can be constructed for a numerical feature.

Features that can be plotted:
- "Revised Cost": True
- "Proposed Units": True
- "Existing Construction Type": True

In [104]:
boxplot_features = ["Revised Cost", "Existing Construction Type", "Proposed Units"]

for feature in boxplot_features:
    print(f"Feature: {feature}, Type: {df[feature].dtype}, Can be plotted: {df[feature].dtype == 'float64' or df[feature].dtype == 'int64'}")

Feature: Revised Cost, Type: float64, Can be plotted: True
Feature: Existing Construction Type, Type: float64, Can be plotted: True
Feature: Proposed Units, Type: float64, Can be plotted: True


### One-hot encoding can be applied to categorical features with a relatively small number of unique values. (will use 5 as a threshold)

Features that can be one-hot encoded:
- "Description": Can be one-hot encoded: False
- "Street Name": Can be one-hot encoded: False
- "Existing Construction Type Description": Can be one-hot encoded: True

In [105]:
ohe_features = ["Description", "Street Name", "Existing Construction Type Description"]

for feature in ohe_features:
    ohe_thld = 5
    unique_values = df[feature].nunique()
    print(f"Feature: {feature}, Unique values: {unique_values}, Can be one-hot encoded: {unique_values <= ohe_thld}")


Feature: Description, Unique values: 134272, Can be one-hot encoded: False
Feature: Street Name, Unique values: 1704, Can be one-hot encoded: False
Feature: Existing Construction Type Description, Unique values: 5, Can be one-hot encoded: True


### Missing values for "Proposed Construction Type" feature can be filled with median.

Median for "Proposed Construction Type" is: `5.0`

In [106]:
median = df["Proposed Construction Type"].median()

print(f"Median value for 'Proposed Construction Type' feature is: {median}")

Median value for 'Proposed Construction Type' feature is: 5.0


### Let's apply the Robust Scaler to "Proposed Units" feature and compare the changes in median value

Original median: `2.0`

RobustScaler transformed median: `0.0`

In [107]:
# Create a RobustScaler instance
robust_scaler = RobustScaler()

proposed_units = df["Proposed Units"].to_numpy().reshape(-1, 1)
proposed_units_scaled = robust_scaler.fit_transform(proposed_units)
# Convert back to a pandas Series for easier analysis
proposed_units_scaled_series = pd.Series(proposed_units_scaled.flatten())
scaled_median = proposed_units_scaled_series.median()

print(f"Original median of 'Proposed Units': {df['Proposed Units'].median()}")
print(f"Median after RobustScaler transformation: {scaled_median}")



Original median of 'Proposed Units': 2.0
Median after RobustScaler transformation: 0.0


### Let's apply the MinMax Scaler to "Existing Construction Type" feature and compare the changes in median value

Original median: `5.0`

MinMaxScaler transformed median: `1.0`

In [108]:
min_max_scaler = MinMaxScaler()

existing_construction_type = df["Existing Construction Type"].to_numpy().reshape(-1, 1)
existing_construction_type_scaled = min_max_scaler.fit_transform(existing_construction_type)
# Convert back to a pandas Series for easier analysis
existing_construction_type_scaled_series = pd.Series(existing_construction_type_scaled.flatten())

original_median = df["Existing Construction Type"].median()
scaled_median = existing_construction_type_scaled_series.median()

print(f"Original median of 'Existing Construction Type': {original_median}")
print(f"Median after MinMaxScaler transformation: {scaled_median}")


Original median of 'Existing Construction Type': 5.0
Median after MinMaxScaler transformation: 1.0


### "Street Name" feature contains `1704` unique values, so it is not good to transform it using OHE method.

In [109]:
unique_street_names = len(df["Street Name"].unique())
print(f"Number of unique street names: {unique_street_names}")


Number of unique street names: 1704


#### The best strategy for dealing with missing values in "Street Number Suffix" column will be completely dropping the column.

*Rationale*:
- The column has almost no information (only 1.11% of data is present)
- Any imputation would be extremely unreliable
- The column likely provides very little predictive value
- Keeping it would mostly introduce noise

In [110]:
missing_values = df["Street Number Suffix"].isna().sum()
missing_percent = (missing_values / len(df)) * 100

print(f"'Street Number Suffix' column has {missing_percent:.2f}% of missing data.")

'Street Number Suffix' column has 98.89% of missing data.


#### The "Existing Use" feature has 93 unique values which do not seem to have any natural ordering, so we should not apply LabelEncoder to it.

In [111]:
unique_uses = df['Existing Use'].unique()
print(unique_uses)

['tourist hotel/motel' 'retail sales' '1 family dwelling' 'apartments' nan
 '2 family dwelling' 'church' 'storage shed' 'office' 'vacant lot'
 'food/beverage hndlng' 'residential hotel' 'filling/service stn'
 'workshop commercial' 'clinics-medic/dental' 'misc group residns.'
 'hospital' 'club' 'barber/beauty salon' 'warehouse,no frnitur' 'school'
 'artist live/work' 'manufacturing' 'garment shops' 'public assmbly other'
 'auto repairs' 'lending institution' 'museum' 'warehouse, furniture'
 'prkng garage/private' 'antenna' 'health studios & gym' 'massage parlor'
 'printing plant' 'parking lot' 'workshop residential' 'power plant'
 'tower' 'mortuary' 'animal sale or care' 'laundry/laundromat' 'nite club'
 'paint store' 'recreation bldg' 'theater' 'prkng garage/public' 'sign'
 'phone xchnge/equip' 'dance hall' 'sfpd or sffd station' 'storage tanks'
 'muni carbarn' 'stadium' 'automobile sales' 'fence/retaining wall'
 'radio & tv stations' 'social care facility' 'amusement center'
 'day car